File: Python/test_money.py
import unittest
from bank import Bank
from money import Money
from portfolio import Portfolio

class TestMoney(unittest.TestCase):

  def setUp(self):
    self.bank = Bank()
    self.bank.add_exchange_rate('EUR', 'USD', 1.2)
    self.bank.add_exchange_rate('USD', 'KRW', 1100)

  def test_multiplication(self):
    tenEuros = Money(10, 'EUR')
    twentyEuros = Money(20, 'EUR')
    self.assertEqual(twentyEuros, tenEuros.times(2))

  def test_division(self):
    originalMoney = Money(4002, 'KRW')
    actualMoneyAfterDivision = originalMoney.divide(4)
    expectedMoneyAfterDivision = Money(1000.5, 'KRW')
    self.assertEqual(expectedMoneyAfterDivision, actualMoneyAfterDivision)

  def test_addition(self):
    fiveDollars = Money(5, 'USD')
    tenDollars = Money(10, 'USD')
    fifteenDollars = Money(15, 'USD')
    portfolio = Portfolio()
    portfolio.add(fiveDollars, tenDollars)
    self.assertEqual(fifteenDollars, portfolio.evaluate(self.bank, 'USD'))

  def test_addition_of_dollars_and_euros(self):
    fiveDollars = Money(5, 'USD')
    tenEuros = Money(10, 'EUR')
    portfolio = Portfolio()
    portfolio.add(fiveDollars, tenEuros)
    expectedValue = Money(17, 'USD')
    actualValue = portfolio.evaluate(self.bank, 'USD')
    self.assertEqual(expectedValue, actualValue, '%s != %s' % (expectedValue, actualValue))

  def test_addition_of_dollars_and_wons(self):
    oneDollar = Money(1, 'USD')
    elevenHundredWons = Money(1100, 'KRW')
    portfolio = Portfolio()
    portfolio.add(oneDollar, elevenHundredWons)
    expectedValue = Money(2200, 'KRW')
    actualValue = portfolio.evaluate(self.bank, 'KRW')
    self.assertEqual(expectedValue, actualValue, '%s != %s' % (expectedValue, actualValue))

  def test_addition_with_multiple_missing_exchange_rates(self):
    oneDollar = Money(1, 'USD')
    oneEuro = Money(1, 'EUR')
    oneWon = Money(1, 'KRW')
    portfolio = Portfolio()
    portfolio.add(oneDollar, oneEuro, oneWon)
    with self.assertRaisesRegex(
      Exception,
      'Missing exchange rate\(s\): \[USD->Kalganid,EUR->Kalganid,KRW->Kalganid\]'
    ):
      portfolio.evaluate(self.bank, 'Kalganid')

  def test_conversion(self):
    bank = Bank()
    bank.add_exchange_rate('EUR', 'USD', 1.2)
    tenEuros = Money(10, 'EUR')
    expectedValue = Money(12, 'USD')
    actualValue = bank.convert(tenEuros, 'USD')
    self.assertEqual(actualValue, expectedValue, '%s != %s' % (expectedValue, actualValue))

  def test_conversion_with_missing_exchange_rate(self):
    bank = Bank()
    tenEuros = Money(10, 'EUR')
    with self.assertRaisesRegex(
      Exception,
      'EUR->Kalganid'
    ):
      bank.convert(tenEuros, 'Kalganid')

if __name__ == '__main__':
  unittest.main()

File: js/test_money.js
const assert = require('assert')
const Money = require('./money')
const Portfolio = require('./portfolio')
const Bank = require('./bank')

class MoneyTest {
  setUp() {
    this.bank = new Bank()
    this.bank.addExchangeRate('EUR', 'USD', 1.2)
    this.bank.addExchangeRate('USD', 'KRW', 1100)
  }

  testMultiplication() {
    let tenEuros = new Money(10, 'EUR')
    let twentyEuros = new Money(20, 'EUR')
    assert.deepStrictEqual(tenEuros.times(2), twentyEuros)
  }

  testDivision() {
    let originalMoney = new Money(4002, 'KRW')
    let actualMoneyAfterDivision = originalMoney.divide(4)
    let expectedMoneyAfterDivision = new Money(1000.5, 'KRW')
    assert.deepStrictEqual(actualMoneyAfterDivision, expectedMoneyAfterDivision)
  }

  testAddition() {
    let fiveDollars = new Money(5, 'USD')
    let tenDollars = new Money(10, 'USD')
    let fifteenDollars = new Money(15, 'USD')
    let portfolio = new Portfolio()
    portfolio.add(fiveDollars, tenDollars)
    assert.deepStrictEqual(
      portfolio.evaluate(new Bank(), 'USD'),
      fifteenDollars
    )
  }

  testAdditionOfDollarsAndEuros() {
    let fiveDollars = new Money(5, 'USD')
    let tenEuros = new Money(10, 'EUR')
    let portfolio = new Portfolio()
    portfolio.add(fiveDollars, tenEuros)
    let expectedValue = new Money(17, 'USD')
    assert.deepStrictEqual(portfolio.evaluate(this.bank, 'USD'), expectedValue)
  }

  testAdditionOfDollarsAndWons() {
    let oneDollar = new Money(1, 'USD')
    let elevenHundredWons = new Money(1100, 'KRW')
    let portfolio = new Portfolio()
    portfolio.add(oneDollar, elevenHundredWons)
    let expectedValue = new Money(2200, 'KRW')
    assert.deepStrictEqual(portfolio.evaluate(this.bank, 'KRW'), expectedValue)
  }

  testAdditionWithMultipleMissingExchangeRates() {
    let oneDollar = new Money(1, 'USD')
    let oneEuro = new Money(1, 'EUR')
    let oneWOn = new Money(1, 'KRW')
    let portfolio = new Portfolio()
    portfolio.add(oneDollar, oneEuro, oneWOn)
    let expectedError = new Error(
      'Missing exchange rate(s): [USD->Kalganid,EUR->Kalganid,KRW->Kalganid]'
    )
    assert.throws(
      () => portfolio.evaluate(new Bank(), 'Kalganid'),
      expectedError
    )
  }

  testCoversion() {
    let tenEuros = new Money(10, 'EUR')
    assert.deepStrictEqual(
      this.bank.convert(tenEuros, 'USD'),
      new Money(12, 'USD')
    )

    this.bank.addExchangeRate('EUR', 'USD', 1.3)
    assert.deepStrictEqual(
      this.bank.convert(tenEuros, 'USD'),
      new Money(13, 'USD')
    )
  }

  testWhatIsTHeConversionRateFromEURToUSD() {
    let tenEuros = new Money(10, 'EUR')
    assert.deepStrictEqual(
      this.bank.convert(tenEuros, 'USD'),
      new Money(12, 'USD')
    )
  }

  testCoversionWithMissingExchangeRates() {
    let bank = new Bank()
    let tenEuros = new Money(10, 'EUR')
    let expectedError = new Error('EUR->Kalganid')
    assert.throws(() => bank.convert(tenEuros, 'Kalganid'), expectedError)
  }

  runAllTests() {
    let testMethods = this.getAllTestMethods()
    testMethods.forEach((m) => {
      console.log(`Running: ${m}()`)
      let method = Reflect.get(this, m)
      try {
        this.setUp()
        Reflect.apply(method, this, [])
      } catch (e) {
        if (e instanceof assert.AssertionError) {
          console.log(e)
        } else {
          throw e
        }
      }
    })
  }

  getAllTestMethods() {
    let moneyPrototype = MoneyTest.prototype
    let allProps = Object.getOwnPropertyNames(moneyPrototype)
    let testMethods = allProps.filter((p) => {
      return typeof moneyPrototype[p] === 'function' && p.startsWith('test')
    })
    return testMethods
  }
}

new MoneyTest().runAllTests()